File: backend/src/main/java/com/project/bearlink/domain/room/service/RoomService.java
package com.project.bearlink.domain.room.service;


import com.project.bearlink.domain.friend.entity.FriendRequest;
import com.project.bearlink.domain.friend.entity.FriendRequestStatus;
import com.project.bearlink.domain.friend.repository.FriendRequestRepository;
import com.project.bearlink.domain.room.dto.*;
import com.project.bearlink.domain.room.entity.InvitationStatus;
import com.project.bearlink.domain.room.entity.LinkRoom;
import com.project.bearlink.domain.room.entity.RoomLink;
import com.project.bearlink.domain.room.entity.RoomMember;
import com.project.bearlink.domain.room.repository.LinkRoomRepository;
import com.project.bearlink.domain.room.repository.RoomLinkRepository;
import com.project.bearlink.domain.room.repository.RoomMemberRepository;

import com.project.bearlink.domain.user.user.entity.User;
import com.project.bearlink.domain.user.user.repository.UserRepository;
import com.project.bearlink.global.exception.ApiException;
import com.project.bearlink.global.exception.ErrorCode;
import lombok.RequiredArgsConstructor;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class RoomService {

    private final UserRepository userRepository;
    private final LinkRoomRepository linkRoomRepository;
    private final RoomMemberRepository roomMemberRepository;
    private final RoomLinkRepository roomLinkRepository;
    private final FriendRequestRepository friendRequestRepository;

    @Transactional(readOnly = false)
    public CreateLinkRoomResponse createRoom(CreateLinkRoomRequest request, User currentUser) {

        LinkRoom room = LinkRoom.builder()
                .name(request.name())
                .owner(currentUser)
                .build();
        linkRoomRepository.save(room);


        RoomMember selfMember = RoomMember.builder()
                .room(room)
                .user(currentUser)
                .status(InvitationStatus.ACCEPTED)
                .invitedAt(LocalDateTime.now())
                .respondedAt(LocalDateTime.now())
                .build();
        roomMemberRepository.save(selfMember);

        return new CreateLinkRoomResponse(room.getId(), room.getName());
    }

    public List<RoomsDto> getRooms(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ApiException(ErrorCode.USER_NOT_FOUND));

        List<LinkRoom> linkRooms = roomMemberRepository.findAcceptedRoomsByUserId(userId);


        return linkRooms.stream()
                .map(room -> new RoomsDto(room.getId(), room.getName()))
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = false)
    public void deleteRooms(Long roomId){
        LinkRoom linkRoom= linkRoomRepository.findById(roomId)
                .orElseThrow(()-> new ApiException(ErrorCode.ROOM_NOT_FOUND));
        List<RoomLink> links = roomLinkRepository.findByRoomId(roomId);

        roomLinkRepository.deleteAll(links);
        linkRoomRepository.delete(linkRoom);
    }

    @Transactional(readOnly = false)
    public void inviteUser(Long roomId, Long userId, User inviter) {
        LinkRoom room = linkRoomRepository.findById(roomId)
                .orElseThrow(() -> new ApiException(ErrorCode.ROOM_NOT_FOUND));
        if (!room.getOwner().getId().equals(inviter.getId())) {
            throw new ApiException(ErrorCode.ROOM_INVITE_FORBIDDEN);
        }

        User invitee = userRepository.findById(userId)
                .orElseThrow(() -> new ApiException(ErrorCode.USER_NOT_FOUND));


        if (roomMemberRepository.existsByRoomAndUser(room, invitee)) {
            throw new ApiException(ErrorCode.ALREADY_INVITED_MEMBER);
        }

        RoomMember member = RoomMember.builder()
                .room(room)
                .user(invitee)
                .status(InvitationStatus.INVITED)
                .invitedAt(LocalDateTime.now())
                .build();
        roomMemberRepository.save(member);
    }

    public List<InvitationResponse> getMyInvitations(User user) {
        List<RoomMember> invitedMembers = roomMemberRepository.findWithRoomByUserAndStatus(user, InvitationStatus.INVITED);

        return invitedMembers.stream()
                .map(m -> new InvitationResponse(
                        m.getId(),
                        m.getRoom().getId(),
                        m.getRoom().getName()
                ))
                .toList();
    }


    @Transactional(readOnly = false)
    public void acceptInvitation(Long roomMemberId, User user) {
        RoomMember member = roomMemberRepository.findById(roomMemberId)
                .orElseThrow(() -> new ApiException(ErrorCode.ROOM_MEMBER_NOT_FOUND));
        if (!member.getUser().getId().equals(user.getId())) {
            throw new ApiException(ErrorCode.INVITATION_ACCESS_DENIED);
        }
        if (member.getStatus() != InvitationStatus.INVITED) {
            throw new ApiException(ErrorCode.INVITATION_ALREADY_ACCEPTED);
        }

        member.setStatus(InvitationStatus.ACCEPTED);
        member.setRespondedAt(LocalDateTime.now());

    }

    @Transactional(readOnly = false)
    public void declineInvitation(Long roomMemberId, User user) {
        RoomMember member = roomMemberRepository.findById(roomMemberId)
                .orElseThrow(() -> new ApiException(ErrorCode.ROOM_MEMBER_NOT_FOUND));
        if (!member.getUser().getId().equals(user.getId())) {
            throw new ApiException(ErrorCode.INVITATION_ACCESS_DENIED);
        }
        if (member.getStatus() != InvitationStatus.INVITED) {
            throw new ApiException(ErrorCode.INVITATION_ALREADY_ACCEPTED);
        }

        member.setStatus(InvitationStatus.DECLINED);
        member.setRespondedAt(LocalDateTime.now());
    }

    public List<RoomLinkListDto> getRoomLinks(Long userId, Long roomId) {
        List<RoomLink> links = roomLinkRepository.findByRoomOwnerIdAndRoomId(userId, roomId);

        return links.stream()
                .map(link -> new RoomLinkListDto(
                        link.getId(),
                        link.getTitle(),
                        link.getUrl()
                ))
                .collect(Collectors.toList());
    }


    public List<InviteFriendWithStatusResponse> getInviteFriendsWithStatus(User currentUser, Long roomId) {
        LinkRoom room = linkRoomRepository.findById(roomId)
                .orElseThrow(() -> new ApiException(ErrorCode.ROOM_NOT_FOUND));

        List<FriendRequest> acceptedFriends = friendRequestRepository.findByRequesterOrReceiverAndStatus(
                currentUser,
                currentUser,
                FriendRequestStatus.ACCEPTED
        );

        List<RoomMember> roomMembers = roomMemberRepository.findByRoomId(roomId);

        Map<Long, InvitationStatus> memberStatusMap = roomMembers.stream()
                .collect(Collectors.toMap(
                        rm -> rm.getUser().getId(),
                        RoomMember::getStatus
                ));

        return acceptedFriends.stream()
                .map(fr -> {
                    User friend = fr.getRequester().getId().equals(currentUser.getId())
                            ? fr.getReceiver()
                            : fr.getRequester();

                    InvitationStatus status = memberStatusMap.getOrDefault(friend.getId(), null);

                    String invitationStatus;
                    if (status == null) {
                        invitationStatus = "NOT_INVITED";
                    } else if (status == InvitationStatus.INVITED) {
                        invitationStatus = "INVITED";
                    } else if (status == InvitationStatus.ACCEPTED) {
                        invitationStatus = "ACCEPTED";
                    } else {
                        invitationStatus = "DECLINED";
                    }

                    return new InviteFriendWithStatusResponse(
                            friend.getId(),
                            friend.getNickname(),
                            friend.getEmail(),
                            invitationStatus
                    );
                })
                .toList();
    }


    public List<RoomMemberList> getMembers(Long roomId){
        LinkRoom room = linkRoomRepository.findById(roomId)
                .orElseThrow(()-> new ApiException(ErrorCode.USER_NOT_FOUND));

        List<RoomMember> members = roomMemberRepository.findByRoomAndStatus(room, InvitationStatus.ACCEPTED);

        return members.stream()
                .map(member -> new RoomMemberList(
                        member.getId(),
                        member.getUser().getNickname()
                ))
                .toList();
    }

    public RoomsDto getRoom(Long roomId){

        LinkRoom linkRoom = linkRoomRepository.findById(roomId)
                .orElseThrow(()->new ApiException(ErrorCode.ROOM_NOT_FOUND));
        return new RoomsDto(
                linkRoom.getId(),
                linkRoom.getName()
        );
    }
}

File: frontend/src/features/room/service/roomService.ts
import { api } from "../../../lib/api";
import {
  CreateLinkRoomRequest,
  CreateLinkRoomResponse,
  RoomInviteRequest,
  RoomsDto,
  RoomLinkListDto,
  RoomInviteListDto,
  InvitationStatus,
  InvitationResponse,
  InviteFriendWithStatusResponse,
  RoomMemberList,
} from "../type/room";
import { RoomMember } from "../type/roomPageTypes";

/**
 * 링크룸 관련 API 서비스
 * 서버의 /api/v1/room 엔드포인트와 통신
 */
export const roomService = {
  /**
   * 링크룸 삭제
   * 서버 엔드포인트: DELETE /api/v1/room?roomId={roomId}
   *
   * @param roomId 삭제할 방 ID
   * @returns void
   */
  deleteRoom: async (roomId: number): Promise<string> => {
    try {
      const response = await api.delete(`/api/v1/room?roomId=${roomId}`);

      // ApiResponse 형태인지 확인하고 처리
      if (response && response.success !== undefined) {
        const apiResponse = response as ApiResponse<string>;
        if (!apiResponse.success) {
          throw new Error(apiResponse.message);
        }
        return apiResponse.message;
      }

      return "방이 삭제되었습니다.";
    } catch (error) {
      throw error;
    }
  },
  /**
   * 새로운 링크룸 생성
   *
   * @param request 방 생성 요청 정보
   * @returns 생성된 방 정보
   */
  async createRoom(
    request: CreateLinkRoomRequest
  ): Promise<CreateLinkRoomResponse> {
    try {
      // Content-Type을 application/json으로 명시적으로 설정
      const options = {
        headers: {
          "Content-Type": "application/json",
        },
      };
      const response = await api.post("/api/v1/room", request, options);

      // ApiResponse 형태인지 확인하고 처리
      if (response && response.success !== undefined) {
        const apiResponse = response as ApiResponse<CreateLinkRoomResponse>;
        if (!apiResponse.success) {
          throw new Error(apiResponse.message);
        }
        return apiResponse.data;
      }

      return response as CreateLinkRoomResponse;
    } catch (error) {
      throw error;
    }
  },

  /**
   * 사용자를 링크룸에 초대
   * 서버 엔드포인트: POST /api/v1/room/{roomId}/invite
   *
   * @param roomId 초대할 방 ID
   * @param userId 초대할 사용자 ID
   * @returns 초대 결과
   */
  async inviteUser(roomId: number, userId: number): Promise<string> {
    try {
      const request: RoomInviteRequest = { userId };
      const response = await api.post(`/api/v1/room/${roomId}/invite`, request);

      // ApiResponse 형태인지 확인하고 처리
      if (response && response.success !== undefined) {
        const apiResponse = response as ApiResponse<string>;
        if (!apiResponse.success) {
          throw new Error(apiResponse.message);
        }
        return apiResponse.message;
      }

      return "초대가 완료되었습니다.";
    } catch (error) {
      throw error;
    }
  },

  /**
   * 현재 사용자의 모든 링크룸 목록 조회
   * 서버 엔드포인트: GET /api/v1/room
   *
   * @returns 링크룸 목록
   */
  async getRooms(): Promise<RoomsDto[]> {
    try {
      const response = await api.get("/api/v1/room");

      // ApiResponse 형태인지 확인하고 처리
      if (response && response.success !== undefined) {
        const apiResponse = response as ApiResponse<RoomsDto[]>;
        if (!apiResponse.success) {
          throw new Error(apiResponse.message);
        }
        return Array.isArray(apiResponse.data) ? apiResponse.data : [];
      }

      return response as RoomsDto[];
    } catch (error) {
      throw error;
    }
  },

  /**
   * 특정 링크룸 조회 (실제 API 연동)
   * 서버 엔드포인트: GET /api/v1/room/{roomId}
   *
   * @param roomId 조회할 방 ID
   * @returns 링크룸 정보
   */
  async getRoomById(roomId: number): Promise<RoomsDto> {
    try {
      const response = await api.get(`/api/v1/room/${roomId}`);

      // ApiResponse 형태인지 확인하고 처리
      if (response && response.success !== undefined) {
        const apiResponse = response as ApiResponse<RoomsDto>;
        if (!apiResponse.success) {
          throw new Error(apiResponse.message);
        }
        return apiResponse.data;
      }

      return response as RoomsDto;
    } catch (error) {
      throw error;
    }
  },

  /**
   * 링크룸 초대 수락
   * 서버 엔드포인트: POST /api/v1/room/invitations/{roomMemberId}/accept
   *
   * @param roomMemberId 수락할 초대의 room member ID
   * @returns void
   */
  async acceptInvitation(roomMemberId: number): Promise<string> {
    try {
      const response = await api.post(
        `/api/v1/room/invitations/${roomMemberId}/accept`
      );

      // ApiResponse 형태인지 확인하고 처리
      if (response && response.success !== undefined) {
        const apiResponse = response as ApiResponse<string>;
        if (!apiResponse.success) {
          throw new Error(apiResponse.message);
        }
        return apiResponse.message;
      }

      return "초대를 수락했습니다.";
    } catch (error) {
      throw error;
    }
  },

  /**
   * 링크룸 초대 거절
   * 서버 엔드포인트: POST /api/v1/room/invitations/{roomMemberId}/decline
   *
   * @param roomMemberId 거절할 초대의 room member ID
   * @returns void
   */
  async declineInvitation(roomMemberId: number): Promise<string> {
    try {
      const response = await api.post(
        `/api/v1/room/invitations/${roomMemberId}/decline`
      );

      // ApiResponse 형태인지 확인하고 처리
      if (response && response.success !== undefined) {
        const apiResponse = response as ApiResponse<string>;
        if (!apiResponse.success) {
          throw new Error(apiResponse.message);
        }
        return apiResponse.message;
      }

      return "초대를 거절했습니다.";
    } catch (error) {
      throw error;
    }
  },

  /**
   * 링크룸 멤버 목록 조회
   * 서버 엔드포인트: GET /api/v1/room/{roomId}/members
   *
   * @param roomId 조회할 방 ID
   * @returns 링크룸 멤버 목록 (RoomMemberList[] 타입)
   */
  async getMembers(roomId: number): Promise<RoomMemberList[]> {
    try {
      const response = await api.get(`/api/v1/room/${roomId}/members`);

      // ApiResponse 형태인지 확인하고 처리
      if (response && response.success !== undefined) {
        const apiResponse = response as ApiResponse<RoomMemberList[]>;
        if (!apiResponse.success) {
          throw new Error(apiResponse.message);
        }
        return Array.isArray(apiResponse.data) ? apiResponse.data : [];
      }

      if (Array.isArray(response)) {
        return response as RoomMemberList[];
      } else {
        return [];
      }
    } catch (error) {
      throw error;
    }
  },

  /**
   * 링크룸에 속한 링크 목록 조회
   * 서버 엔드포인트: GET /api/v1/room/links?roomId={roomId}
   *
   * @param roomId 조회할 방 ID
   * @returns 링크룸의 링크 목록
   */
  async getRoomLinks(roomId: number): Promise<RoomLinkListDto[]> {
    try {
      const response = await api.get(`/api/v1/room/links?roomId=${roomId}`);

      // ApiResponse 형태인지 확인하고 처리
      if (response && response.success !== undefined) {
        const apiResponse = response as ApiResponse<RoomLinkListDto[]>;
        if (!apiResponse.success) {
          throw new Error(apiResponse.message);
        }
        return Array.isArray(apiResponse.data) ? apiResponse.data : [];
      }

      return response as RoomLinkListDto[];
    } catch (error) {
      throw error;
    }
  },

  /**
   * 초대 가능한 친구 목록 조회 (상태 포함)
   * 서버 엔드포인트: GET /api/v1/room/{roomId}/invite-friends
   *
   * @param roomId 조회할 방 ID
   * @returns 초대할 수 있는 친구 목록 (상태 정보 포함)
   */
  async getInvitableFriends(
    roomId: number
  ): Promise<InviteFriendWithStatusResponse[]> {
    try {
      // 캐시를 무시하고 최신 데이터를 가져오기 위한 랜덤 쿼리 파라미터 추가
      const timestamp = new Date().getTime();
      const response = await api.get(
        `/api/v1/room/${roomId}/invite-friends?_t=${timestamp}`
      );

      // ApiResponse 형태인지 확인하고 처리
      if (response && response.success !== undefined) {
        const apiResponse = response as ApiResponse<
          InviteFriendWithStatusResponse[]
        >;
        if (!apiResponse.success) {
          throw new Error(apiResponse.message);
        }
        return Array.isArray(apiResponse.data) ? apiResponse.data : [];
      }

      // 응답 데이터가 유효한지 확인
      if (Array.isArray(response)) {
        return response as InviteFriendWithStatusResponse[];
      } else {
        return [];
      }
    } catch (error) {
      throw error; // 에러를 그대로 전달하여 호출한 쪽에서 처리하도록 함
    }
  },

  /**
   * 사용자가 받은 초대 목록 조회
   * 서버 엔드포인트: GET /api/v1/room/invitations
   *
   * @returns 사용자가 받은 초대 목록 (InvitationResponse 배열)
   */
  async getMyInvitations(): Promise<InvitationResponse[]> {
    try {
      const response = await api.get("/api/v1/room/invitations");

      // ApiResponse 형태인지 확인하고 처리
      if (response && response.success !== undefined) {
        const apiResponse = response as ApiResponse<InvitationResponse[]>;
        if (!apiResponse.success) {
          throw new Error(apiResponse.message);
        }
        return Array.isArray(apiResponse.data) ? apiResponse.data : [];
      }

      // 응답 형식 검증
      if (Array.isArray(response)) {
        return response as InvitationResponse[];
      } else {
        return [];
      }
    } catch (error: any) {
      throw error;
    }
  },
};

// 편의를 위한 alias
export default roomService;
